### DISCUSSION TITLE
Newbie struggling with mesh setup

### DISCUSSION THREAD

**Discussion by piranha32** [thliebig/openEMS-Project#130] [Q&A Usage]
I'm completely new to EM modelling, and since I need to design and build a discone antenna, I used it as an opportunity to learn openEMS.
Based on the examples and tutorials, I created the antenna model, set up simulation, and I can generate simple graphs for S11 and antenna impedance. The excitation port is connected between the center of the top metal disk, and the center of a metal disk capping the top end of the cone. One thing that I can't get over is the setup of the simulation mesh. 

As a starting point, I use antenna dimensions generated by an online calculator, and I'd expect to see from the model results roughly matching the requirements. Unfortunately, what I get is nowhere close to my expectations (not very surprising, the results from the calculator may be off), and drastically changes with even small changes to the mesh structure.

I got to using mesh with resolution of lambda_min/30 for the volume enclosing the antenna (with margins), and lambda_min/20 for the rest of the simulation volume (with 3/4 of max dimension of margin along each of the axes), and I still can't obtain trustworthy-looking results.

I would appreciate if you could help me improve the experiment setup.
Attached is a python program with full simulation code.

[disconelab.txt](https://github.com/thliebig/openEMS-Project/files/12973327/disconelab.txt)

**Comment by Sarajevo67** [comment]
Unfortunately, openEMS isn't best suited for simulation of curved 3D shapes. I tried to simulate thick wire helical antenna, but only for the purpose of visualisation of currents and fields. No trustworthy impedance simulation could be expected (as stated by @thliebig ).
My solution was to use NEC MoM simulation for helix impedance and openEMS for 2D shapes, like MSL impedance matching transformer.

**Reply by piranha32** [reply]
After some more experiments, I managed to get to a stable solution with max mesh size reduced to 1/3 of the length of the gap between the cone and the disk, which is smaller than lambda_min. In the hindsight, it makes perfect sense.
The only downside is that the generated fine mesh is big, and the simulation takes between 20 min and an hour (surprisingly, using even finer mesh shortens the simulation time).

Is there a way to generate one smooth mesh between x1 and x2, and then another, with different resolution between x3 and x4? So far, I only figured out how to create overlying meshes, by expanding the meshed area.

In my first attempt I tried to use cylindrical coordinates to build the model, but I could not figure out how to properly build a cone. AddRotPoly creates a geometry which can be show in CSXCAD, but with big, fat warning that "CSPrimPolygon can not be defined in non Cartesian coordinate systems", and the "result may be unexpected".

**Reply by Sarajevo67** [reply]
Can you share your discone dimensions and simulation impedance results?

**Reply by piranha32** [reply]
Here are the dimensions (lengths in mm):
```
disc_dia: 75
disc_thickness: 1
cone_upper_dia: 4.2
cone_lower_dia: 107
cone_angle: 60
cone_thickness: 0.8

disc_cone_distance: 1.2

feed_R: 50

f_min: 800e6
f_max: 6e9
```

And here is the feed point impedance plot for fine mesh generated with max res of `disc_cone_distance/5` in z direction, and `min(C0/(f_max)/unit/30, cone_upper_dia/5)` in x and y directions.
![discone imp 5](https://github.com/thliebig/openEMS-Project/assets/680825/d26f26ad-cd63-47a9-b7b0-edf15771c3e2)

**Reply by Sarajevo67** [reply]
You took too wide frequency range for port results. Reduce it to 0.7-3.4GHz to remove that deep reactance drop.
Use discone antenna calculator from [this site](https://3g-aerial.biz/en/online-calculations/antenna-calculations/discone-antenna-online-calculator) There you will find ANSYS simulation results for this discone with Fmin=800MHz. Compare them with openEMS.

Best Regards,

DD

**Reply by piranha32** [reply]
I used calculator from https://www.changpuak.ch/electronics/DisconeAntennaCalculator.php, and the dimensions came out almost identical. Both calculators look so similar, that I think they may be based on the same code.
Thanks for the link to the ANSYS analysis. I'll compare the results with my simulation. I got to the point where the simulation results don't change when I refine the mesh even further, so hopefully I've got it right.

As for the too wide frequency range, is this too wide for the simulation or for the antenna?

**Reply by Sarajevo67** [reply]
Calculator references Fmin=800 Fmax=3400 MHz for antenna. For simulation excitation I would use Fc 2.5GHz with 2 GHz from each side (0.5-4.5 GHz)

**Reply by glompos21** [reply]
After some testing i found the discussion. As of now (25/12/2024) does the opemEMS capable of simulating a 3d shape? 
I try to simulate an asymmetrical bionical antenna (for start will be a symmetrical one).
I used Freecad for the design, the macro to generate the python code. I have put it on the following [GitHub repo](https://github.com/glompos21/asymetric_biconical). 
Am at the right road to get some results or i have to find another software?

**Reply by thliebig** [reply]
openEMS should be able to simulate this. The only issue is to create a mesh that does recognize the curved shapes correctly.
Ideally the wall thickness is larger than the mesh size. Otherwise the PEC (metal) surface is not closed.

At some point in the future openEMS needs a feature to map a metallic surface safely to the FDTD grid.
I recommend to use the --debug-PEC option...
